ovs-service/src/main/java/org/dcsa/ovs/util/pagination/Paginator.java

```diff
@@ -0,0 +1,131 @@
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.net.URLEncoder;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Helper class for doing "cursor" based pagination using spring-data.
+ */
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class Paginator {
+  private final ObjectMapper objectMapper;
+
+  /**
+   * Parses the incoming request and returns a Cursor for querying jpa repositories.
+   */
+  public Cursor parseRequest(HttpServletRequest request, CursorDefaults cursorDefaults) {
+    String cursor = request.getParameter("cursor");
+    if (cursor != null) {
+      return cursorFromString(cursor);
+    }
+
+    int pageSize = cursorDefaults.getPageSize();
+    String limit = request.getParameter("limit");
+    if (limit != null) {
+      pageSize = Integer.parseInt(limit);
+    }
+
+    return new Cursor(0, pageSize, cursorDefaults.getSortBy());
+  }
+
+  /**
+   * Sets DCSA pagination headers.
+   *
+   * @param totalPages total pages in a result (see also org.springframework.data.domain.Page)
+   */
+  public void setPageHeaders(
+    HttpServletRequest request,
+    HttpServletResponse response,
+    Cursor currentCursor,
+    int totalPages,
+    String... allowedParameters
+  ) {
+    Set<String> allowedParametersSet = allowedParameters.length == 0 ? null : Arrays.stream(allowedParameters).collect(Collectors.toSet());
+    setPageHeaders(request, response, currentCursor, totalPages, allowedParametersSet);
+  }
+
+  /**
+   * Sets DCSA pagination headers.
+   *
+   * @param totalPages total pages in a result (see also org.springframework.data.domain.Page)
+   */
+  public void setPageHeaders(
+    HttpServletRequest request,
+    HttpServletResponse response,
+    Cursor currentCursor,
+    int totalPages,
+    Set<String> allowedParameters
+  ) {
+    String basePath = request.getScheme() + "://" +
+      request.getServerName() + ":" + request.getServerPort() +
+      request.getRequestURI() + "?cursor=";
+    String parameters = getRequestParameters(request, allowedParameters);
+
+    response.setHeader("Current-Page", basePath + cursorToString(currentCursor) + parameters);
+    response.setHeader("First-Page", basePath + cursorToString(currentCursor.withPage(0)) + parameters);
+
+    if (currentCursor.getPage() > 0) {
+      response.setHeader("Previous-Page", basePath + cursorToString(currentCursor.withRelativePage(-1)) + parameters);
+    }
+    if (totalPages > currentCursor.getPage() + 1) {
+      response.setHeader("Next-Page", basePath + cursorToString(currentCursor.withRelativePage(1)) + parameters);
+    }
+  }
+
+  @SneakyThrows
+  @VisibleForTesting
+  String cursorToString(Cursor cursor) {
+    return urlEncode(base64Encode(objectMapper.writeValueAsBytes(cursor)));
```

Comment:
urlEncode and base64Encode can be consolidated into java.util.Base64.getUrlDecoder()